pydicom/fileset.py

```diff
@@ -0,0 +1,2720 @@
+                ds.file_meta.TransferSyntaxUID
+            )
+
+            yield record
+            return
+
+        # Four-level records: PATIENT -> STUDY -> SERIES -> leaf
+        records = []
+        leaf_type = _four_level_record_type(ds)
+        for record_type in ["PATIENT", "STUDY", "SERIES", leaf_type]:
+            try:
+                record = DIRECTORY_RECORDERS[record_type](ds)
+            except ValueError as exc:
+                raise ValueError(
+                    f"Unable to use the default pydicom '{record_type}' "
+                    "record creator as the instance is missing a "
+                    "required element or value. Either update the instance "
+                    "or use 'FileSet.add_custom()' instead"
+                ) from exc
+
+            record.OffsetOfTheNextDirectoryRecord = 0
+            record.RecordInUseFlag = 0xFFFF
+            record.OffsetOfReferencedLowerLevelDirectoryEntity = 0
+            record.DirectoryRecordType = record_type
+            if "SpecificCharacterSet" in ds:
+                record.SpecificCharacterSet = ds.SpecificCharacterSet
+
+            records.append(record)
+
+        # Add the instance referencing elements to the leaf
+        leaf = records[3]
+        leaf.ReferencedFileID = None
+        leaf.ReferencedSOPClassUIDInFile = ds.SOPClassUID
+        leaf.ReferencedSOPInstanceUIDInFile = ds.SOPInstanceUID
+        leaf.ReferencedTransferSyntaxUIDInFile = (
+            ds.file_meta.TransferSyntaxUID
+        )
+
+        yield from records
+
+    def remove(
+        self, instance: Union[FileInstance, List[FileInstance]]
+    ) -> None:
+        """Stage instance(s) for removal from the File-set.
+
+        Parameters
+        ----------
+        instance : pydicom.fileset.FileInstance or a list of FileInstance
+            The instance(s) to remove from the File-set.
+        """
+        if isinstance(instance, list):
+            for item in instance:
+                self.remove(item)
+            return
+
+        if instance not in self._instances:
+            raise ValueError("No such instance in the File-set")
+
+        # If staged for addition, no longer add
+        if instance.SOPInstanceUID in self._stage['+']:
+            leaf = instance.node
+            del leaf.parent[leaf]
+            del self._stage['+'][instance.SOPInstanceUID]
+            # Delete file from stage
+            try:
+                instance._stage_path.unlink()
+            except FileNotFoundError:
+                pass
+            instance._apply_stage('-')
+            self._instances.remove(instance)
+
+        # Stage for removal if not already done
+        elif instance.SOPInstanceUID not in self._stage['-']:
+            instance._apply_stage('-')
+            self._stage['-'][instance.SOPInstanceUID] = instance
+            self._instances.remove(instance)
+
+    # TODO: Fix str representation
+    def __str__(self) -> str:
```

Comment:
Is this TODO still relevant?

Reply:
Yeah, still working on it. It's almost done

Reply:
Should be done now